app/controllers/agent/WhatToDoController.scala

```diff
@@ -53,20 +56,50 @@ class WhatToDoController @Inject()(val messagesApi: MessagesApi,
   }
 
 
-  def submit(name: String, nonMTDfB: Boolean): Action[AnyContent] = authenticate {
+  def submit: Action[AnyContent] = authenticate.async {
     implicit user =>
+
       if (appConfig.features.whereToGoFeature()) {
-        WhatToDoForm.whatToDoForm.bindFromRequest().fold(
-          error => BadRequest(views.html.agent.whatToDo(error, name, nonMTDfB)),
-          data => data.value match {
-            case SubmitReturn.value => Redirect(appConfig.returnDeadlinesUrl)
-            case ViewReturn.value => Redirect(appConfig.submittedReturnsUrl(DateTime.now(DateTimeZone.UTC).year().get()))
-            case ChangeDetails.value => Redirect(appConfig.manageVatCustomerDetailsUrl)
-            case ViewCertificate.value => Redirect(appConfig.vatCertificateUrl)
+        (user.session.get(SessionKeys.mtdVatAgentClientName), user.session.get(SessionKeys.mtdVatAgentMandationStatus)) match {
+          case (Some(clientName), Some(mandationStatus)) =>
+            Future.successful(WhatToDoForm.whatToDoForm.bindFromRequest().fold(
+              error => BadRequest(views.html.agent.whatToDo(error, clientName, mandationStatus == nonMTDfB)),
+              data => {
+                data.value match {
+                  case SubmitReturn.value => Redirect(appConfig.returnDeadlinesUrl)
+                  case ViewReturn.value => Redirect(appConfig.submittedReturnsUrl(DateTime.now(DateTimeZone.UTC).year().get()))
+                  case ChangeDetails.value => Redirect(appConfig.manageVatCustomerDetailsUrl)
+                  case ViewCertificate.value => Redirect(appConfig.vatCertificateUrl)
+                }
+              }
+            ).removingFromSession(SessionKeys.mtdVatAgentClientName, SessionKeys.mtdVatAgentMandationStatus))
+          case _ => {
+            Logger.debug("[WhatToDoController][submit] - unable to get at least one of mtdVatAgentClientName, mtdVatAgentMandationStatus from session")
+            detailsCallAndSubmit
           }
-        )
+        }
       } else {
-        serviceErrorHandler.showInternalServerError
+        Future.successful(serviceErrorHandler.showInternalServerError)
       }
   }
+
+  private def detailsCallAndSubmit(implicit user: User[_]): Future[Result] = {
+    customerDetailsService.getCustomerDetails(user.vrn).map {
+      case Right(details) =>
```

Comment:
A lot of this logic is duplicated and doesn't need to be. You could refactor this method to only construct the 'BadRequest' view by either using session values or calling vat subscription again, e.g.
```
private def badRequestResult(error: Form[WhatToDoModel])(implicit user: User[_]): Future[Result] = {
    (user.session.get(SessionKeys.mtdVatAgentClientName), user.session.get(SessionKeys.mtdVatAgentMandationStatus)) match {
      case (Some(clientName), Some(mandationStatus)) =>
        Future.successful(BadRequest(views.html.agent.whatToDo(error, clientName, mandationStatus == nonMTDfB)))
      case _ =>
        customerDetailsService.getCustomerDetails(user.vrn).map {
          case Right(details) => BadRequest(views.html.agent.whatToDo(error, details.clientName, details.mandationStatus == nonMTDfB))
          case Left(error) =>
            Logger.warn(s"[WhatToDoController][submit] - received an error from CustomerDetailsService: $error")
            serviceErrorHandler.showInternalServerError
      }
    }
  }
```

The `submit` method could therefore be changed to call the private method whenever there are errors in the form e.g.
```
def submit: Action[AnyContent] = authenticate.async {
    implicit user =>
      if (appConfig.features.whereToGoFeature()) {
        WhatToDoForm.whatToDoForm.bindFromRequest().fold(
          error => badRequestResult(error),
          data => Future.successful(
            data.value match {
              case SubmitReturn.value => Redirect(appConfig.returnDeadlinesUrl)
              case ViewReturn.value => Redirect(appConfig.submittedReturnsUrl(DateTime.now(DateTimeZone.UTC).year().get()))
              case ChangeDetails.value => Redirect(appConfig.manageVatCustomerDetailsUrl)
              case ViewCertificate.value => Redirect(appConfig.vatCertificateUrl)
            }
          )
        )
      } else {
        Future.successful(serviceErrorHandler.showInternalServerError)
      }
  }
```
I haven't ran the tests but in theory it would separate concerns a bit nicer

Reply:
Remember to add and remove to session - I didn't include in above example :P